pkg/controllers/management/globalnamespacerbac/rbac_common.go

```diff
@@ -0,0 +1,225 @@
+package globalnamespacerbac
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/rancher/rancher/pkg/namespace"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	rbacv1 "github.com/rancher/types/apis/rbac.authorization.k8s.io/v1"
+	"github.com/rancher/types/config"
+
+	k8srbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	allAccess                 = "all"
+	updateAccess              = "update"
+	readOnlyAccess            = "readonly"
+	MultiClusterAppResource   = "multiclusterapps"
+	GlobalDNSResource         = "globaldnses"
+	GlobalDNSProviderResource = "globaldnsproviders"
+	CreatorIDAnn              = "field.cattle.io/creatorId"
+)
+
+func CreateRoleAndRoleBinding(resource string, name string, members []v3.Member, creatorID string, managementContext *config.ManagementContext) error {
+	/* Create 3 Roles containing the resource (multiclusterapp or globalDNS), and the current multiclusterapp/globalDNS in resourceNames list
+	1. Role with all verbs (everything access, includes creator); name multiclusterapp.Name + "-ma" / (globalDNS.Name + "-ga")
+	2. Role with "get", "list" "watch" verbs (ReadOnly access); name multiclusterapp.Name + "-mr" / (globalDNS.Name + "-gr")
+	3. Role with "update" verb (Upgrade access); name multiclusterapp.Name + "-mu" / (globalDNS.Name + "-gu")
+	*/
+	accessTypes := []string{allAccess, updateAccess, readOnlyAccess}
+	for _, r := range accessTypes {
+		if _, err := createRole(resource, r, name, managementContext); err != nil {
+			return err
+		}
+	}
+
+	// Create a roleBinding referring the role with everything access, and containing creator of the multiclusterapp, along with
+	// any members that have everything access
+	var allAccessSubjects, readOnlyAccessSubjects, updateAccessSubjects []k8srbacv1.Subject
+	allAccessSubjects = append(allAccessSubjects, k8srbacv1.Subject{Kind: "User", Name: creatorID, APIGroup: rbacv1.GroupName})
+	for _, m := range members {
+		s, err := buildSubjectForMember(m, managementContext)
+		if err != nil {
+			return err
+		}
+		switch m.AccessType {
+		case allAccess:
+			allAccessSubjects = append(allAccessSubjects, s)
+		case updateAccess:
+			updateAccessSubjects = append(updateAccessSubjects, s)
+		case readOnlyAccess:
+			readOnlyAccessSubjects = append(readOnlyAccessSubjects, s)
+		default:
+			readOnlyAccessSubjects = append(readOnlyAccessSubjects, s)
+		}
+	}
+
+	if _, err := createRoleBinding(allAccess, name, allAccessSubjects, managementContext, resource); err != nil {
```

Comment:
I think we should sort the subjects (by user name). Otherwise deep equal might always return false